demos/custom-tabs-example-app/src/main/java/org/chromium/customtabsdemos/EngagementSignalsActivity.java

```diff
@@ -0,0 +1,193 @@
+
+import android.os.RemoteException;
+import android.util.Log;
+import android.view.View;
+import android.widget.EditText;
+import android.widget.TextView;
+
+import androidx.browser.customtabs.CustomTabsCallback;
+import androidx.browser.customtabs.CustomTabsClient;
+import androidx.browser.customtabs.CustomTabsIntent;
+import androidx.browser.customtabs.CustomTabsSession;
+import androidx.browser.customtabs.EngagementSignalsCallback;
+
+public class EngagementSignalsActivity extends AppCompatActivity implements View.OnClickListener {
+    private static final String TAG = EngagementSignalsActivity.class.getSimpleName();
+    private static final int INITIAL_HEIGHT_DEFAULT_PX = 600;
+
+    private EditText mUrlEditText;
+    private TextView mTextVerticalScroll;
+    private TextView mTextGreatestPercentage;
+    private TextView mTextSessionEnd;
+    private TextView mTextNavigation;
+
+    private ServiceConnection mConnection;
+    private CustomTabsClient mClient;
+    private CustomTabsSession mCustomTabsSession;
+
+    private ServiceConnectionCallback mServiceConnectionCallback = new ServiceConnectionCallback() {
+        @Override
+        public void onServiceConnected(CustomTabsClient client) {
+            mClient = client;
+            mCustomTabsSession = mClient.newSession(mCustomTabsCallback);
+            try {
+                boolean engagementSignalsApiAvailable = mCustomTabsSession.isEngagementSignalsApiAvailable(Bundle.EMPTY);
+                if (!engagementSignalsApiAvailable) {
+                    Log.d(TAG, "CustomTab Engagement signals not available, make sure to use the " +
+                            "latest Chrome version and enable via chrome://flags/#cct-real-time-engagement-signals");
+                    return;
+                }
+                boolean signalsCallback = mCustomTabsSession.setEngagementSignalsCallback(mEngagementSignalsCallback, Bundle.EMPTY);
+                if (!signalsCallback) {
+                    Log.w(TAG, "Could not set EngagementSignalsCallback");
+                }
+            } catch (RemoteException e) {
+                Log.w(TAG, "The Service died while responding to the request.", e);
+            } catch (UnsupportedOperationException e) {
+                Log.w(TAG, "Engagement Signals API isn't supported by the browser.", e);
+            }
+        }
+
+        @Override
+        public void onServiceDisconnected() {
+            mClient = null;
+            mConnection = null;
+            mCustomTabsSession = null;
+        }
+    };
+
+    private EngagementSignalsCallback mEngagementSignalsCallback = new EngagementSignalsCallback() {
+        @Override
+        public void onVerticalScrollEvent(boolean isDirectionUp, @NonNull Bundle extras) {
+            Log.d(TAG, "onVerticalScrollEvent (isDirectionUp=" + isDirectionUp + ')');
+            mTextVerticalScroll.setText("vertical scroll " + (isDirectionUp ? "UP️" : "DOWN"));
```

Comment:
Would it be better to append the text, instead of overwriting it? That way users will see a log of all of their actions? As it is, I assume if someone opened a CCT then scrolled down, they could miss the vertical scroll events and just see scroll percentage increased events.

(This isn't that critical, just a quality of life improvement.)

Reply:
Thought about this as well, I'm using a partial custom tab so it's harder to display the log due to limited screen space. I think it works well as is.

Reply:
OK then.